test/test_quantization.py

```diff
@@ -0,0 +1,254 @@
+
+        tq.convert_to_quantized(myModel)
+        self.assertEqual(hasattr(myModel, 'quant'), False)
+        self.assertEqual(hasattr(myModel.fc2, 'quant'), True)
+        self.assertEqual(hasattr(myModel.fc2, 'observer'), False)
+        self.assertEqual(type(myModel.fc1), torch.nn.Linear)
+        self.assertEqual(type(myModel.fc2), nnq.Linear)
+        default_eval_fn(myModel, calib_data)
+
+    def test_nested1(self):
+        myModel = NestedModel()
+        qConfigDict = {
+            'fc3': default_qconfig,
+            'sub2.fc2': default_qconfig
+        }
+        tq.prepare(myModel, qConfigDict)
+        self.assertEqual(hasattr(myModel, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub1, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub1.fc, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub2, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub2.fc1, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub2.fc2, 'quant'), True)
+        self.assertEqual(hasattr(myModel.sub2.fc2, 'observer'), True)
+        self.assertEqual(hasattr(myModel.fc3, 'quant'), True)
+        self.assertEqual(hasattr(myModel.fc3, 'observer'), True)
+
+        default_eval_fn(myModel, calib_data)
+
+        tq.convert_to_quantized(myModel)
+        self.assertEqual(hasattr(myModel, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub1, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub1.fc, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub2, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub2.fc1, 'quant'), False)
+        self.assertEqual(hasattr(myModel.sub2.fc2, 'quant'), True)
+        self.assertEqual(hasattr(myModel.sub2.fc2, 'observer'), False)
+        self.assertEqual(hasattr(myModel.fc3, 'quant'), True)
+        self.assertEqual(hasattr(myModel.fc3, 'observer'), False)
+        self.assertEqual(type(myModel.sub1.fc), torch.nn.Linear)
+        self.assertEqual(type(myModel.sub2.fc1), torch.nn.Linear)
+        self.assertEqual(type(myModel.sub2.fc2), nnq.Linear)
+        self.assertEqual(type(myModel.fc3), nnq.Linear)
+        default_eval_fn(myModel, calib_data)
+
+    def test_nested2(self):
+        """
+        If we add quant dequant for the whole module then we can eliminate
+        extra quant dequant between modules. This is what current implementation
+        supports.
+        However, a more complete support would make test_nested3 work as well,
+        but still keep the current behavior. That is to say, when user provides
+        configurations for finer grained modules, we operate on that level, e.g.
+        if user have a key 'sub2.fc2', then we don't treat 'sub2' as a terminal
+        module, instead we'll operate on the same level as 'sub2.fc2'.
+        """
+        myModel = NestedModel()
+        qConfigDict = {
+            'fc3': default_qconfig,
+            'sub2': default_qconfig
```

Comment:
I'm still a bit uneasy about using these settings to control non-inserting quant/dequant. There are a few problems with it:

- It's non-obvious that `forward` of the module needs to be runnable with quantized values. It's very easy to miss some functional call there and have wtf moment
- the property that particular module can be wrapped as quant/dequant is a property of module **type** and not module **instance**. So it'd be more natural to have a separate configuration that lists which models support "entire module" quantization or just enable it for people who want to do it manually
- I'd interpret qConfig applied to a parent module (e.g. "sub2") to be just observer settings to be applied to all modules in the subgraph that are supported and **skip** module types that are unknown. Otherwise we're signing ourselves up for spelling every individual sub-module explicitly if I just want to get behavior similar to torch/jit/quantized.py which is swapping out all modules that support quantization.

Here's a different proposal - we have two types of settings:
- qconfig that describes observer settings only
- list of module types that can be quantized whole-sale. By default they are the keys from DEFAULT_MODULE_MAPPING list but I one can add new settings.

Then .prepare would receive both of those setting and apply quant/dequant insertion only for modules of supported types who are in the parts of the tree present in qconfig.

The graph mode api would just receive qconfig directly as the compiler knows best which modules are "supported" and where to insert quantization blocks (though if we're thinking of customization, we might pass some per-module-type settings too).

Thoughts?

Reply:
Good points,  I agree with some of them:
1. Great point about forward. We have no way of knowing if a module's forward consists of quantizable operations, short of doing tracing. So the default module swapping behavior should replace a module with q->quant-module->dq or fq->quant-module->fq in the QAT case. If the user is sure that a given module is fully quantizable, then we could have a 'strict=True' option that would not insert quant dequants. This will not be the default.
2. I think having quantization controllable at the level of a module instance is important. Think of resnet-101, in this case we have 101 layers that are identical, (nn.conv2d), but we do want to be able to control which layers we want to quantize. It is also important for the automatic precision selection work as we can figure out which parts of a network we want to quantize.
3. Agreed with (3). My understanding of (3) is that if I have a module for which I have a qconfig, I would attempt to quantize every sub-module within it. If I find a submodule that is not supported (say a softmax), that would not be modified. Note that using strict=True would dump an error message in this case.
4. We are planning on a quantization debugger (proposal is a quip doc) that allows for whitelist/blacklist) of ops etc as a bootcamp/HAM task.
For graph mode, the API would remain the same, the user would get the benefits of fusion (better insertion of quant/dequant), no extra quant/dequants (as we know the entire execution graph).

Reply:
2. I'm fully agreeing that we want to have control of quantization on module instance level. I was talking only about inserting quant/dequant. So in ResNet101 case I can still specify qconfig on pieces of modules. But if I want to say that module MyAwesomeModule has a good "forward" such that quantizing everything inside it is sufficient to just put quant/dequant around it - we should specify it by passing 'MyAwesomeModule''s type, not instance as it's safe to apply to all instances that are covered in qconfig

3. Agree with your reasoning. But as outlined above I'd decouple qconfig for quant/dequant insertion a bit

More examples of what exactly I mean. Let's say we have a hierarchy:
```
m : MyModel
  sub1 : MyModule
    fc : FC
    conv: Conv
  sub2 : MyModule
    fc : FC
    conv: Conv
  sub3 : MyOtherModule
    fc1 : FC
    fc2 : FC
```

Let's say MyModule has good forward and MyOtherModule has a "bad" one.

Then `qconfig={"": ...}` would quantize all individual FC/CONV layers and insert quant/dequant around each of them.

If I pass `qconfig={"": ...}, whole_modules={MyModule}`, then quant/dequant would be inserted around `sub1, sub2, sub3.fc1, sub3.fc2`

If I pass `qconfig={"sub2": ...}, whole_modules={MyModule}`, then quant/dequant would be inserted around `sub2` only (sub1 is not in qconfig).

If I pass `qconfig={"sub2": default_qconfig, "sub2.fc": None}, whole_modules={MyModule}`, then quant/dequant would be inserted around `sub2.conv` only - because part of MyModule is not quantized according to qconfig.

Does it make sense?

Also for more complex quant/dequant insertions in eager mode - we should really just ask people to manually insert QuantStub modules into them.

Reply:
I agree, this makes more sense than the 'strict=True' that I was proposing.  I was only defining default behaviors in the previous comments: 
- Default would correspond to whole_modules={} and we would insert q-dq per module that we can convert. 
- If whole_module is specified, we would error out if MyModule had something within that is not quantizable at preparation time itself, If there is an operation in forward that is not quantizable (say a sqrt for example), the error would be seen only at run-time as the assumption of the whole_module being quantizable is violated.

Reply:
makes sense, I'll add module setting.

Reply:
maybe a different PR, this is out for too long.